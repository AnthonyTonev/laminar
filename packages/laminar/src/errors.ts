import { LoggerMetadata } from './logger';
import { PgError } from './pg/pg-error';

/**
 * Get metadata object for loggers from an error instance
 */
export const toErrorMetadata = (error: Error): LoggerMetadata => {
  if (error instanceof PgError) {
    return { queryText: error.queryText, position: error.position, stack: error.stack };
  } else {
    return { stack: error.stack };
  }
};
