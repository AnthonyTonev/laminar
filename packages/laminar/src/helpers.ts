export const toArray = (value: unknown): unknown[] =>
  Array.isArray(value) ? value : value ? [value] : [];

type Obj = Record<string, unknown>;

const isObj = (obj: unknown): obj is Obj => typeof obj === 'object' && obj !== null;

const setQuery = (path: string[], value: unknown, obj: Obj): Obj | unknown[] => {
  const [current, ...rest] = path;
  if (current) {
    const currentValue = obj[current];
    return {
      ...obj,
      [current]: rest.length
        ? setQuery(rest, value, isObj(currentValue) ? currentValue : {})
        : value,
    };
  } else {
    return toArray(rest.length ? setQuery(rest, value, obj) : value);
  }
};

const toQueryPath = (key: string): string[] => key.replace(/\]/g, '').split('[');

export const parseQueryObjects = (plain: Obj): Obj =>
  Object.entries(plain)
    .map<[string, unknown]>(([key, val]) => [
      key,
      typeof val === 'string' && val.includes(',') ? val.split(',') : val,
    ])
    .reduce((all, [key, val]) => setQuery(toQueryPath(key), val, all), {});
