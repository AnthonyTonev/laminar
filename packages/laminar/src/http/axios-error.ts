import { inspect } from 'util';

/**
 * An axios error for easier testing.
 *
 * @category http
 */
export class AxiosError extends Error {
  public constructor(
    public readonly code: number,
    public readonly message: string,
    public readonly data: unknown,
    public readonly stack?: string,
  ) {
    super(`HttpError ${code}\nMessage: ${message}\nBody: ${inspect(data, { depth: 10 })}\n`);
  }
}
