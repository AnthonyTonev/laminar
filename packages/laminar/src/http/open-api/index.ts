export { openApi, defaultError } from './open-api';
export { securityOk, isSecurityOk, securityError, securityRedirect } from './security';
export {
  OapiContext,
  AppRouteOapi,
  OapiPath,
  OapiAuthInfo,
  SecurityOk,
  Security,
  RequestSecurityResolver,
  OapiSecurityResolver,
  OapiSecurity,
  ResponseOapi,
  OapiPaths,
  OapiConfig,
} from './types';
