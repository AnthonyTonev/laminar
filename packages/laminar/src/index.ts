export { laminar } from './laminar';
export { request } from './request';
export {
  response,
  isResponse,
  message,
  redirect,
  file,
  toResponse,
  extendResponse,
} from './response';
export { withCors, CorsConfig } from './middleware/cors';
export { withResponseTime, ResponseTimeConfig } from './middleware/responseTime';
export { withLogger, Logger, LoggerContext } from './middleware/logger';
export { HttpError } from './HttpError';
export { selectRoute, toRoute, router, get, post, patch, del, options, put } from './router';
export {
  Addition,
  Context,
  Laminar,
  LaminarCookie,
  LaminarRequest,
  LaminarResponse,
  RouteResolver,
  Method,
  Middleware,
  Resolver,
  ResolverResponse,
  Route,
  RouteContext,
} from './types';
