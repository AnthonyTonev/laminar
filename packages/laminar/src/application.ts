import { AxiosError } from './http/axios-error';
import { LoggerLike } from './logger';
import { Service } from './types';

/**
 * A recursive list of {@link Service} instances or arrays of {@link Service} instances.
 * @category application
 */
export type InitOrder = Service | InitOrder[];

/**
 * An instance of an application. Should hold the `initOrder` for all of the {@link Service}-es of the application.
 * Would be started in the `initOrder` (and stopped in the reverse of `initOrder` order).
 *
 * Arrays in `initOrder` define services that need to be started / stopped in parallel.
 * @category application
 */
export interface Application {
  /**
   * A recursive nested list of {@link Service}.
   */
  initOrder: InitOrder[];
  /**
   * A logger object that would be used to print out the progress of starting / stopping things.
   * You can pass an {@link Service} here as well, and it would be stopped last.
   */
  logger?: LoggerLike | (LoggerLike & Service);
}

/**
 * @category application
 */
export interface StartLogger {
  /**
   * Should start the logger service, if its a {@link Service} used to break services recursion.
   */
  startLogger?: boolean;
}

/**
 * @category application
 */
export interface StopLogger {
  /**
   * Should stop the logger service, if its a {@link Service} used to break services recursion.
   */
  stopLogger?: boolean;
}

const startService = async (item: Service, logger?: LoggerLike | (LoggerLike & Service)): Promise<void> => {
  logger?.info(`⏫ Starting ${item.describe()}`);
  await item.start();
  logger?.info(`✅ Started ${item.describe()}`);
};

const stopService = async (item: Service, logger?: LoggerLike | (LoggerLike & Service)): Promise<void> => {
  logger?.info(`⏬ Stopping ${item.describe()}`);
  await item.stop();
  logger?.info(`❎ Stopped ${item.describe()}`);
};

/**
 * Start all the {@link Service} objects in `initOrder` sequentially. If it has an array of {@link Service}, will start them in parallel.
 *
 * ```typescript
 * // Will start s1, s2 and s3 in that sequential order
 * start({ initOrder: [s1, s2, s3] });
 *
 * // Will start s1 and s2, then start s3, s4 and s5 in parallel, and lastly start s6
 * start({ initOrder: [s1, s2, [s3, s4, s5], s6] });
 *```
 * @category application
 */
export async function start({ initOrder, logger, startLogger = true }: Application & StartLogger): Promise<void> {
  if (logger && 'start' in logger && startLogger) {
    await logger.start();
  }

  for (const item of initOrder) {
    await ('start' in item
      ? startService(item, logger)
      : Promise.all(
          item.map((child) =>
            'start' in child ? startService(child, logger) : start({ initOrder: child, logger, startLogger: false }),
          ),
        ));
  }
}

/**
 * Stop all the {@link Service} objects in `initOrder` sequentially, in reverse order from {@link start}
 *
 * ```typescript
 * // Will stop s3, s2 and s1 in that sequential order
 * stop({ initOrder: [s1, s2, s3] });
 *
 * // Will stop s6, then s3, s4 and s5 in parallel, and lastly will stop s2 and s1 sequentially
 * start({ initOrder: [s1, s2, [s3, s4, s5], s6] });
 *```
 * @category application
 */
export async function stop({ initOrder, logger, stopLogger = true }: Application & StopLogger): Promise<void> {
  for (const item of initOrder.reverse()) {
    await ('stop' in item
      ? stopService(item, logger)
      : Promise.all(
          item.map((child) =>
            'stop' in child ? stopService(child, logger) : stop({ initOrder: child, logger, stopLogger: false }),
          ),
        ));
  }
  if (logger && 'stop' in logger && stopLogger) {
    logger?.info(`❎ Stop ${logger.describe()}`);
    await logger.stop();
  }
}

/**
 * Run {@link start}, and add a listener to SIGTERM, that would call {@link stop}
 *
 * @category application
 */
export async function init<TApplication extends Application>(app: TApplication): Promise<TApplication> {
  await start(app);
  process.once('SIGTERM', () => stop(app));
  return app;
}

/**
 * Run {@link start}, execute the predicate, and when its finished, call {@link stop}.
 * Useful for testing out the application as you can put your test code in the predicate
 *
 * @category application
 */
export async function run(app: Application, predicate: (app: Application) => Promise<void>): Promise<void> {
  await start(app);
  try {
    await predicate(app);
  } catch (error) {
    throw error.response ? new AxiosError(error.response.status, error.response.data, error.stack) : error;
  } finally {
    await stop(app);
  }
}
