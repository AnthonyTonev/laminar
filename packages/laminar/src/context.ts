import * as cookie from 'cookie';
import { IncomingMessage } from 'http';
import { parse } from 'url';
import { parseQueryObjects } from './helpers';
import { Method, Context } from './types';

export const toContext = async (req: IncomingMessage): Promise<Context> => {
  const url = req.url ? parse(req.url, true) : { query: {} };
  return {
    body: req,
    url,
    method: req.method as Method,
    query: parseQueryObjects(url.query),
    cookies: req.headers.cookie ? cookie.parse(req.headers.cookie) : undefined,
    headers: req.headers,
  };
};
