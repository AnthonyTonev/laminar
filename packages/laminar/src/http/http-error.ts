import { OutgoingHttpHeaders } from 'http';

/**
 * A generic exception.
 *
 * When thrown, the default laminar error handler will convert it to a response object, using the code for status and headers and body appropriately.
 *
 * @category http
 */
export class HttpError extends Error {
  public readonly code: number;
  public readonly body: { message: string; [key: string]: unknown };
  public readonly headers?: OutgoingHttpHeaders;
  public readonly stack?: string;

  public constructor(
    code: number,
    body: { message: string; [key: string]: unknown },
    headers?: OutgoingHttpHeaders,
    stack?: string,
  ) {
    super(body.message);
    this.code = code;
    this.body = body;
    this.headers = headers;
    this.stack = stack;
  }
}

/**
 * Check if a response from {@link OapiSecurityResolver} is a {@link HttpError} object, indicating a failed security check
 *
 * @category http
 */
export function isHttpError(item: unknown): item is HttpError {
  return typeof item === 'object' && item !== null && 'code' in item && 'body' in item;
}
