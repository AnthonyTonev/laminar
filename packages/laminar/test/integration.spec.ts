import axios from 'axios';
import { Server } from 'http';
import {
  del,
  get,
  laminar,
  options,
  patch,
  post,
  put,
  redirect,
  response,
  router,
  message,
  createResponseTime,
  Middleware,
} from '../src';
import { createLogging, createBodyParser } from '../src';
import { promisify } from 'util';

let server: Server;

describe('Integration', () => {
  afterEach(() => promisify(server.close.bind(server))());

  it('Should process response', async () => {
    const loggerMock = { log: jest.fn() };
    const logging = createLogging(loggerMock);
    const bodyParser = createBodyParser();
    const responseTime = createResponseTime();

    interface DBContext {
      getUser: (id: string) => string | undefined;
      delUser: (id: string) => void;
      setUser: (id: string, name: string) => void;
    }

    const db: Middleware<DBContext> = next => {
      const users: { [key: string]: string } = {
        10: 'John',
        20: 'Tom',
      };

      const dbCtx: DBContext = {
        getUser: id => users[id],
        setUser: (id, name) => {
          users[id] = name;
        },
        delUser: id => {
          delete users[id];
        },
      };

      return ctx => {
        return next({ ...ctx, ...dbCtx });
      };
    };

    server = await laminar({
      port: 8050,
      app: bodyParser(
        responseTime(
          db(
            logging(
              router(
                get('/.well-known/health-check', () => ({ health: 'ok' })),
                get('/link', () => redirect('http://localhost:8050/destination')),
                get('/link-other', () =>
                  redirect('http://localhost:8050/destination', {
                    headers: { Authorization: 'Bearer 123' },
                  }),
                ),
                get('/destination', () => ({ arrived: true })),
                get('/error', () => {
                  throw new Error('unknown');
                }),
                options('/users/{id}', () =>
                  response({
                    headers: {
                      'Access-Control-Allow-Origin': 'http://localhost:8050',
                      'Access-Control-Allow-Methods': 'GET,POST,DELETE',
                    },
                  }),
                ),
                get('/users/{id}', ({ path, logger, getUser }) => {
                  logger.log('debug', `Getting id ${path.id}`);
                  const user = getUser(path.id);

                  if (user) {
                    return Promise.resolve({ id: path.id, name: user });
                  } else {
                    return message(404, { message: 'No User Found' });
                  }
                }),
                put('/users', ({ body, logger, setUser }) => {
                  logger.log('debug', `Test Body ${body.name}`);
                  setUser(body.id, body.name);
                  return { added: true };
                }),
                patch('/users/{id}', ({ path, body, getUser, setUser }) => {
                  const user = getUser(path.id);
                  if (user) {
                    setUser(body.id, body.name);
                    return { patched: true };
                  } else {
                    return message(404, { message: 'No User Found' });
                  }
                }),
                post('/users/{id}', ({ path, body, getUser, setUser }) => {
                  const user = getUser(path.id);
                  if (user) {
                    setUser(path.id, body.name);
                    return { saved: true };
                  } else {
                    return message(404, { message: 'No User Found' });
                  }
                }),
                del('/users/{id}', ({ path, getUser, delUser }) => {
                  const user = getUser(path.id);
                  if (user) {
                    delUser(path.id);
                    return { deleted: true };
                  } else {
                    return message(404, { message: 'No User Found' });
                  }
                }),
              ),
            ),
          ),
        ),
      ),
    });

    const api = axios.create({ baseURL: 'http://localhost:8050' });

    await expect(api.get('/unknown-url')).rejects.toHaveProperty(
      'response',
      expect.objectContaining({
        status: 404,
        data: { message: 'Path GET /unknown-url not found' },
      }),
    );

    await expect(api.get('/error')).rejects.toHaveProperty(
      'response',
      expect.objectContaining({
        status: 500,
        data: { message: 'unknown' },
      }),
    );

    await expect(api.get('/.well-known/health-check')).resolves.toMatchObject({
      status: 200,
      data: { health: 'ok' },
    });

    await expect(api.get('/.well-known/health-check/')).resolves.toMatchObject({
      status: 200,
      data: { health: 'ok' },
    });

    await expect(api.get('/.well-known/health-check/other')).rejects.toHaveProperty(
      'response',
      expect.objectContaining({
        status: 404,
        data: { message: 'Path GET /.well-known/health-check/other not found' },
      }),
    );

    await expect(api.get('/link')).resolves.toMatchObject({
      status: 200,
      data: { arrived: true },
    });

    await expect(api.get('/link-other')).resolves.toMatchObject({
      status: 200,
      data: { arrived: true },
    });

    await expect(api.request({ url: '/users/10', method: 'OPTIONS' })).resolves.toMatchObject({
      status: 200,
      headers: expect.objectContaining({
        'access-control-allow-methods': 'GET,POST,DELETE',
        'access-control-allow-origin': 'http://localhost:8050',
      }),
    });

    await expect(api.get('/users/10')).resolves.toMatchObject({
      status: 200,
      data: { id: '10', name: 'John' },
    });

    await expect(api.get('/users/20')).resolves.toMatchObject({
      status: 200,
      data: { id: '20', name: 'Tom' },
    });

    await expect(api.get('/users/30')).rejects.toHaveProperty(
      'response',
      expect.objectContaining({
        status: 404,
        data: { message: 'No User Found' },
      }),
    );

    await expect(api.post('/users/10', { name: 'Kostas' })).resolves.toMatchObject({
      status: 200,
      data: { saved: true },
    });

    await expect(api.patch('/users/20', { name: 'Pathing' })).resolves.toMatchObject({
      status: 200,
      data: { patched: true },
    });

    await expect(api.get('/users/10')).resolves.toMatchObject({
      status: 200,
      data: { id: '10', name: 'Kostas' },
    });

    await expect(api.delete('/users/10')).resolves.toMatchObject({
      status: 200,
      data: { deleted: true },
    });

    await expect(api.get('/users/10')).rejects.toHaveProperty(
      'response',
      expect.objectContaining({
        status: 404,
        data: { message: 'No User Found' },
      }),
    );

    await expect(api.put('/users', { id: 30, name: 'Added' })).resolves.toMatchObject({
      status: 200,
      data: { added: true },
    });

    await expect(api.get('/users/30')).resolves.toMatchObject({
      status: 200,
      data: { id: '30', name: 'Added' },
    });

    const expectedLog = [
      ['info', 'Request', { uri: 'GET /unknown-url', body: '[Stream]' }],
      ['info', 'Response', { status: 404, body: { message: 'Path GET /unknown-url not found' } }],
      ['info', 'Request', { uri: 'GET /error', body: '[Stream]' }],
      ['error', 'Error', { message: 'unknown', stack: expect.any(String) }],
      ['info', 'Request', { uri: 'GET /.well-known/health-check', body: '[Stream]' }],
      ['info', 'Response', { status: 200, body: { health: 'ok' } }],
      ['info', 'Request', { uri: 'GET /.well-known/health-check/', body: '[Stream]' }],
      ['info', 'Response', { status: 200, body: { health: 'ok' } }],
      ['info', 'Request', { uri: 'GET /.well-known/health-check/other', body: '[Stream]' }],
      [
        'info',
        'Response',
        { status: 404, body: { message: 'Path GET /.well-known/health-check/other not found' } },
      ],
      ['info', 'Request', { uri: 'GET /link', body: '[Stream]' }],
      [
        'info',
        'Response',
        { status: 302, body: 'Redirecting to http://localhost:8050/destination.' },
      ],
      ['info', 'Request', { uri: 'GET /destination', body: '[Stream]' }],
      ['info', 'Response', { status: 200, body: { arrived: true } }],
      ['info', 'Request', { uri: 'GET /link-other', body: '[Stream]' }],
      [
        'info',
        'Response',
        { status: 302, body: 'Redirecting to http://localhost:8050/destination.' },
      ],
      ['info', 'Request', { uri: 'GET /destination', body: '[Stream]' }],
      ['info', 'Response', { status: 200, body: { arrived: true } }],
      ['info', 'Request', { uri: 'OPTIONS /users/10', body: '[Stream]' }],
      ['info', 'Response', { status: 200, body: undefined }],
      ['info', 'Request', { uri: 'GET /users/10', body: '[Stream]' }],
      ['debug', 'Getting id 10'],
      ['info', 'Response', { status: 200, body: { id: '10', name: 'John' } }],
      ['info', 'Request', { uri: 'GET /users/20', body: '[Stream]' }],
      ['debug', 'Getting id 20'],
      ['info', 'Response', { status: 200, body: { id: '20', name: 'Tom' } }],
      ['info', 'Request', { uri: 'GET /users/30', body: '[Stream]' }],
      ['debug', 'Getting id 30'],
      ['info', 'Response', { status: 404, body: { message: 'No User Found' } }],
      ['info', 'Request', { uri: 'POST /users/10', body: { name: 'Kostas' } }],
      ['info', 'Response', { status: 200, body: { saved: true } }],
      ['info', 'Request', { uri: 'PATCH /users/20', body: { name: 'Pathing' } }],
      ['info', 'Response', { status: 200, body: { patched: true } }],
      ['info', 'Request', { uri: 'GET /users/10', body: '[Stream]' }],
      ['debug', 'Getting id 10'],
      ['info', 'Response', { status: 200, body: { id: '10', name: 'Kostas' } }],
      ['info', 'Request', { uri: 'DELETE /users/10', body: '[Stream]' }],
      ['info', 'Response', { status: 200, body: { deleted: true } }],
      ['info', 'Request', { uri: 'GET /users/10', body: '[Stream]' }],
      ['debug', 'Getting id 10'],
      ['info', 'Response', { status: 404, body: { message: 'No User Found' } }],
      ['info', 'Request', { uri: 'PUT /users', body: { id: 30, name: 'Added' } }],
      ['debug', 'Test Body Added'],
      ['info', 'Response', { status: 200, body: { added: true } }],
      ['info', 'Request', { uri: 'GET /users/30', body: '[Stream]' }],
      ['debug', 'Getting id 30'],
      ['info', 'Response', { status: 200, body: { id: '30', name: 'Added' } }],
    ];

    expectedLog.forEach((item, index) => {
      expect(loggerMock.log).toHaveBeenNthCalledWith(index + 1, ...item);
    });
  });
});
