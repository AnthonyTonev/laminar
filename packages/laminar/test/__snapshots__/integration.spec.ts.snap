// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Integration Should process response 1`] = `
Array [
  Array [
    "error",
    "Path GET /unknown-url not found",
    Object {
      "error": [Error: Path GET /unknown-url not found],
      "method": "GET",
      "path": "/unknown-url",
    },
  ],
  Array [
    "error",
    "unknown",
    Object {
      "error": [Error: unknown],
      "method": "GET",
      "path": "/error",
    },
  ],
  Array [
    "error",
    "Path GET /.well-known/health-check/other not found",
    Object {
      "error": [Error: Path GET /.well-known/health-check/other not found],
      "method": "GET",
      "path": "/.well-known/health-check/other",
    },
  ],
  Array [
    "debug",
    "Getting id 10",
  ],
  Array [
    "debug",
    "Getting id 20",
  ],
  Array [
    "debug",
    "Getting id 30",
  ],
  Array [
    "error",
    "No User Found",
    Object {
      "error": [Error: No User Found],
      "method": "GET",
      "path": "/users/30",
    },
  ],
  Array [
    "debug",
    "Getting id 10",
  ],
  Array [
    "debug",
    "Getting id 10",
  ],
  Array [
    "error",
    "No User Found",
    Object {
      "error": [Error: No User Found],
      "method": "GET",
      "path": "/users/10",
    },
  ],
  Array [
    "debug",
    "Test Body Added",
  ],
  Array [
    "debug",
    "Getting id 30",
  ],
]
`;
